README.md
# Social Credit Bot
## **A Bot for Social Credit Score management**
\
A **disnake** bot made by [Calamity](https://clmty.xyz) and [KrutosX](https://github.com/KrutosVIP) within the rise of the Social Credit meme, and now you can have moderators manage your members' Social Credit score on your own server!

![the chinese emoji thumbs up](https://images-ext-2.discordapp.net/external/ca1_bYpaa-gjPJvExilq5G1zgaj1OJLhj5OCSzkxdvI/%3Fsize%3D1024/https/cdn.discordapp.com/avatars/895688121285509151/a89b1630ae57e50586c078a5f496828b.webp?width=676&height=676)

---

## **Official Bot: [Invite](https://discord.com/oauth2/authorize?client_id=895688121285509151&permissions=380104993856&scope=bot)**
<br>

---
## How to run your own instance?

0. **do not remove the mentions of original creators**.

1. You will need to create an empty **credit.json** file in the root folder of the bot.

Just in case: fill it like this:
```json
{

}
```

**THIS IS IMPORTANT: MAKE SURE THAT THE BOT ISN'T ON ANY SERVERS WHEN RESETTING THE DB. THIS WILL CAUSE A LOT OF PROBLEMS.**
**THIS IS IMPORTANT: MAKE SURE THAT THE BOT ISN'T ON ANY SERVERS WHEN RESETTING THE DB. THIS WILL CAUSE A LOT OF PROBLEMS.**
**THIS IS IMPORTANT: MAKE SURE THAT THE BOT ISN'T ON ANY SERVERS WHEN RESETTING THE DB. THIS WILL CAUSE A LOT OF PROBLEMS.**

*in case of updating the bot: backup the credit.json file using `sc.backup`, then copy it back from `backup/timedate/credit.json`*

2. You will need to create a **token.json** file in the root folder of the bot.

The file should look like this:
```json
{
    "vendor.token": "BOT TOKEN GOES HERE"
}
```

3. Go to **env.json** and make sure that the bot type is set to `release` to not have debug logs in the console.

4. Just launch the bot using `~$ python bot.py`

Our bot has an advanced logging system, so it's gonna tell you in case anything goes wrong.

You should be up and running at this point!
<br><br>

---
## Contributing?
### **Yes, please!**

Fork the repo, implement the feature, submit a PR. Simple as that.
<br><br>

---
## Support?
**If you have any problems or questions: [join our Discord Server](https://discord.gg/RNSSJatNbq)**, and ask a question in `#sc-support`. We mainly offer support in Russian, but we can also answer in English.
<br><br>

---
## Anything else?
Open an issue, a PR, ask something in [our Discord](https://discord.gg/RNSSJatNbq), don't be shy.


CHANGELOG
[.gitignore] Added __pycache__ #Krutos
[bot.py] [cogs/credit.py] Added logs type - [W]arning, [E]rror... #Krutos
[functions.py] Added file to delete some shitcode. #Krutos
[functions.py] debug_info function is now here. #Krutos
[bot.py] Importing functions.py to output our debug_info. #Krutos
[CHANGELOG] Just created changelog? -w- #Krutos
[bot.py] If cogs.credit unloaded, print out warning to console. #Krutos
[functions.py] Some functions migrated here. Also, created std replacement. And, backup function is also here. #Krutos
[bot.py] Added logout #Krutos
[env.json] Add enviroment file. #Krutos
[functions.py] Added RAM info into Debug info. #Krutos
[bot.py] Added universal cog loader. #Krutos
[functions.py] [bot.py] Add version check #Krutos
[env.json] Change keys #Krutos
[bot.py] Now handle CommandNotFound #Krutos
[credit.py] Add limits to social credit #Calamity
[utils.py] Add cooldown to ping # Krutos
[credit.py] Add command to load db, sc is unique for any guild # Krutos
[utils.py] Fixed guild leave.

[everything] mostly did a big update by myself # Calamity

[credit.py] replaced id2 with gid (guild id) # Calamity 
[credit.py] implemented server reset # Calamity
[bot.py] members intent # Calamity
[bot.py] guild count # Calamity

[holy fuck] no idea, but i patched this thing to work for the message content thing # Calamity

cogs/kek.py
import disnake, json
from disnake.ext import commands
from datetime import datetime
from colorama import Back, Fore, Style

class kek(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    def load(self):
        with open("credit.json", "r", encoding="utf-8") as file:
            self.bot.db = json.load(file)

    def check_lang(self, guild):
        self.load()
        gid = str(guild.id)
        return self.bot.db[gid]["lang"]

    @commands.command(aliases=['tiananmen','тяньаньмэнь','тианамен','танки','площадь','square','протест','1989','тяньмэнь','тиананмен','protest']) 
    async def tanks(self, ctx):
        lang = self.check_lang(ctx.guild)
        if lang == "RU":
            emb = disnake.Embed(
                title='Ничего не происходить в 1989 на Площади Тяньаньмэнь!',
                color = ctx.guild.me.color
            ) # ничего не происходить в комментарий строка 27 файл kek.py
        elif lang == "EN":
            emb = disnake.Embed(
                title='Nothing happened in 1989 on the Tiananmen Square!',
                color = ctx.guild.me.color
            )
        emb.set_image(url='https://media.discordapp.net/attachments/883778578318258222/897894172915294239/unknown.png')
        emb.set_footer(text=ctx.author,icon_url=ctx.author.avatar.url)
        await ctx.reply(embed=emb)

def setup(bot):
    bot.add_cog(kek(bot))

functions.py
import sys, os, disnake, shutil, psutil, json
from colorama import Fore, Back, Style
from datetime import datetime
from copy import copy
if not os.path.isdir("logs"): os.mkdir("logs")

def version_check(major, minor):
    if sys.version_info.major != major:
        print(f'{Fore.RED}[{datetime.now()}] [C] [VCHECK] - Minimal version to boot is {major}.{minor}, but your is {sys.version_info.major}.{sys.version_info.minor}!{Style.RESET_ALL}')
        sys.exit()
    elif sys.version_info.minor < minor:
            print(f'{Fore.RED}[{datetime.now()}] [C] [VCHECK] - Minimal version to boot is {major}.{minor}, but your is {sys.version_info.major}.{sys.version_info.minor}!{Style.RESET_ALL}')
            sys.exit()

    print(f'{Fore.LIGHTCYAN_EX}[{datetime.now()}] [C] [VCHECK] - Version Check passed - {major}.{minor} >= {sys.version_info.major}.{sys.version_info.minor}{Style.RESET_ALL}')


def backup():
    backup = os.path.join("backup", datetime.now().isoformat().replace(":", "."))
    os.mkdir(backup)
    shutil.copyfile("credit.json", os.path.join(backup, "credit.json"))
    print(f'{Fore.LIGHTGREEN_EX}[{datetime.now()}] [I] [BACKUP] - Backed up database to "{backup}"!{Style.RESET_ALL}')
    del backup

class STD: # sexually transmitted disease
    def __init__(self, date):
        self.basesys = sys.stdout
        self.stdout = []
        self.log = os.path.join("logs", f"log-stdout-{date}.log")

    def flush(self, *args):
        self.basesys.flush()
    
    def write(self, *args):
        self.stdout.append(*args)
        self.basesys.write(*args)
        with open(self.log, "w", encoding="utf-8") as f:
            f.write("".join(self.stdout)) # Метод офкорс я из пита скопироваль)

class STDerr:
    def __init__(self, date):
        self.basesys = sys.stderr
        self.stderr = []
        self.log = os.path.join("logs", f"log-stderr-{date}.log")

    def flush(self, *args):
        self.basesys.flush()
    
    def write(self, *args):
        self.stderr.append(*args)
        self.basesys.write(*args)
        with open(self.log, "w", encoding="utf-8") as f:
            f.write("".join(self.stderr)) # Метод офкорс я из пита скопироваль) x2

def redefine_std():
    date = datetime.now().isoformat().replace(":", ".")
    sys.stdout = STD(date)
    sys.stderr = STDerr(date)

def release_info(conf):
    string= "\n -> Bot in Release mode. Version: "+ conf["ro.bot.version"] +"\n -> Python Core " + sys.version + "\n"
    impl = sys.implementation
    impl = f"{impl.name}/Tag: {impl.cache_tag}"
    string += f" -> Implementation: {impl}"

    print(f'{Fore.LIGHTGREEN_EX}{Style.BRIGHT}[{datetime.now()}] [D] [RELINF] - Release info: {string}{Style.RESET_ALL}')
    del string, impl

def debug_info(conf):
    def count_percent(percent, all):
        return round(100/(all/percent)) 
    def get_progress(percent):
        blocks = round(percent / 10)
        not_busy_blocks = 10-blocks
        progress = (blocks * "█" * 2) + (not_busy_blocks * "░" *2)
        return progress, str(percent) + "%"

    string= "\n -> Python Core " + sys.version + "\n"
    if sys.platform in ["win32", "cygwin"]:
        platform = "Windows "
        ver = sys.getwindowsversion()
        platform += f"{ver.major}(.{ver.minor}) build {ver.build}"
        string+= f" -> Platform: {platform} \n"
        del platform, ver
    impl = sys.implementation
    impl = f"{impl.name}/Tag: {impl.cache_tag}"
    string += f" -> Implementation: {impl} \n"
    string += f" -> Core C API Version: {sys.api_version} \n"

    mem = psutil.virtual_memory()
    percent = count_percent(mem.used, mem.total)
    string += f' -> VRam: {" ".join(get_progress(percent))} \n'
    string += f" -> env.json Info: \n"
    string += f" --> Bot Version (ro.bot.version): {conf['ro.bot.version']} \n" 
    string += f" --> LogD (ro.bot.logd): {conf['ro.bot.logd']} \n" 
    string += f" --> Build Type (ro.build.type): {conf['ro.build.type']} \n" 
    string += f" --> Python Mininal Version (ro.python.minimal): {conf['ro.python.minimal']}" 

    print(f'{Fore.LIGHTGREEN_EX}{Style.BRIGHT}[{datetime.now()}] [D] [DEBINF] - Debug info: {string}{Style.RESET_ALL}')
    del string, impl

# ==========
# = ОШИБКИ =
# ==========

async def perms_error(ctx):
    emb = disnake.Embed(
        title='Permissions Error!',
        description='The CCP didn\'t gave you the **Manage Roles** permission, so you cannot edit the Social Credit Scores! Lots Of Laughs! `(англ. язык временно, потом фикс)`',
        color=0xff0000
    )
    emb.set_image(url='https://media.discordapp.net/attachments/883778578783821865/896453128185081896/unknown.png')
    emb.set_footer(text=ctx.author,icon_url=ctx.author.avatar.url)
    await ctx.send(embed=emb)

async def member_not_found(ctx, type):
    emb = disnake.Embed(
        title='Citizen not found!',
        description=f'The Bot couldn\'t find the citizen for editing their Social Credit Score! `(англ. язык временно, потом фикс)`',
        color=0xff0000
    )
    emb.set_image(url='https://media.discordapp.net/attachments/883779765415337995/896458794886918225/unknown.png')
    emb.set_footer(text=ctx.author, icon_url=ctx.author.avatar.url)
    await ctx.send(embed=emb)

cogs/utils.py
import disnake, colorama, sys, os, json
from datetime import datetime
from colorama import Back, Fore, Style
from disnake.ext import commands, tasks
from functions import backup
from time import sleep

class Utils(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    def load(self):
        with open("credit.json", "r", encoding="utf-8") as file:
            self.bot.db = json.load(file)

    def check_lang(self, guild):
        self.load()
        gid = str(guild.id)
        return self.bot.db[gid]["lang"]

    @commands.Cog.listener()
    async def on_guild_join(self, guild):
        print(f"{Fore.LIGHTGREEN_EX}[{datetime.now()}] [I] [+GUILD] - Bot joined a new guild: '{guild.name}'. Member count: '{guild.member_count}'.{Style.RESET_ALL}")
        emb = disnake.Embed(
            title=':wave: Здравствуйте!',
            description="Я есть Социальный Кредит бот разработанный `Calamity#3483` и `KrutosX#3599` совместно с Коммунистическая Партия Китай.",
            color=0xb8493c
        )
        emb.add_field(
            name=":key: Что я уметь?",
            value="Данный бот уметь добавление/отнимание/установка Социальный Кредит, и Китай Пропаганда (молчание!).\nУзнать больше: `sc.help`",
            inline=False
        )
        emb.add_field(
            name=":flag_gb: English?",
            value="Switch to the English language using `sc.lang EN`. (only availible for admins)"
        )
        for i in guild.text_channels:
            perm = i.permissions_for(guild.me)
            if perm.send_messages:
                await i.send(embed=emb)
                break

    @commands.Cog.listener()
    async def on_guild_remove(self, guild):
        print(f"{Fore.LIGHTRED_EX}[{datetime.now()}] [I] [-GUILD] - Bot left guild: '{guild.name}'. Member count: '{guild.member_count}'.{Style.RESET_ALL}")
    
    @commands.cooldown(rate=1, per=10)
    @commands.command(description="пингануть")
    async def ping(self, ctx):
        lang = self.check_lang(ctx.guild)
        if lang == "RU":
            emb = disnake.Embed(
                title='Понг!',
                description=f'Пинг: {int(round(self.bot.latency, 4) * 1000)}',
                color=ctx.guild.me.color
            )
        elif lang == "EN":
            emb = disnake.Embed(
                title='Pong!',
                description=f'Latency: {int(round(self.bot.latency, 4) * 1000)}',
                color=ctx.guild.me.color
            )
        await ctx.send(embed=emb,ephemeral=True)
        print(f"{Fore.LIGHTCYAN_EX}[{datetime.now()}] [I] [COMMND] - 'ping' command executed by {ctx.author}.{Style.RESET_ALL}")

    @commands.command(description="информация о боте")
    async def version(self, ctx):
        lang = self.check_lang(ctx.guild)
        if lang == "RU":
            emb = disnake.Embed(
                title='Версии',
                description=f'Версия: `{self.bot.conf["ro.bot.version"]}`\nВерсия Python: `{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}`\nВерсия disnake: `{disnake.__version__}`',
                color=0xff0000
            )
        elif lang == "EN":
            emb = disnake.Embed(
                title='Versions',
                description=f'Bot Version: `{self.bot.conf["ro.bot.version"]}`\nPython Version: `{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}`\ndisnake Version: `{disnake.__version__}`',
                color=0xff0000
            )
        await ctx.send(embed=emb)
        print(f"{Fore.LIGHTCYAN_EX}[{datetime.now()}] [I] [COMMND] - 'version' command executed by {ctx.author}.{Style.RESET_ALL}")

    @commands.command(aliases=["quit", 'logout', 'выйти', 'выключить', 'вырубить', 'poweroff'])
    async def logoff(self, ctx):
        if self.bot.ownercheck(ctx.author.id):
            await ctx.reply("Logging off...")
            # 4(x^2) - 3x + 1 = 0
            # a = 4, b = -3, c = +3
            # D = b^2 - 4ac = -3^2 - 4 * 4 * 3 = 9 - 48 = -39
            # 3(x^2) + 7x − 6 = 0
            # D = b^2 - 4ac = 7^2 - 4 * 3 * (-6) = 49 - 12*(-6) = 49 + 72 = 121
            # D>0; 2k.
            # x1 = (-b + sqrt(D))/2a
            # x2 =  
            # 🂠 back
            # 🂱🂲🂳🂴🂵🂶🂷🂸🂹🂺🂻🂽🂾 hearts
            # 🃁🃂🃃🃄🃅🃆🃇🃈🃉🃊🃋🃍🃎 diamonds
            # 🃑🃒🃓🃔🃕🃖🃗🃘🃙🃚🃛🃝🃞 spades
            # 🂡🂢🂣🂤🂥🂦🂧🂨🂩🂪🂫🂭🂮 clubs
            # 25x - 17 = 4x - 5 - 13x + 14 + 34x
            # 25x - 17 = 38x - 13x + 9
            # 38x - 17 != 38x + 9
            # Корней нет, иди нахуй. - Каламя
            # Я хоть и знаю, что это не мне, но я не против! - Крутос
            # Делаем дз с Крутосом епта - 11.10.2021, 22:15
            print(f"{Fore.LIGHTCYAN_EX}[{datetime.now()}] [I] [COMMND] - 'logoff' command executed by {ctx.author}.{Style.RESET_ALL}")
            backup()
            if 'credit' in self.bot.cogs: 
                self.bot.cogs['credit'].save()
            print(f"{Fore.LIGHTCYAN_EX}[{datetime.now()}] [I] [COMMND] - Saved and backed up database to do safe logout.{Style.RESET_ALL}")
            await self.bot.close()

            print(f"{Fore.LIGHTCYAN_EX}[{datetime.now()}] [I] [CLIENT] - Logged out.{Style.RESET_ALL}")
            sys.exit()

    @commands.command()
    async def reload(self, ctx):
        if self.bot.ownercheck(ctx.author.id):
            for i in os.listdir("cogs"):
                if os.path.isfile(os.path.join("cogs", i)):
                    if f'cogs.{i[:-3]}' in self.bot.extensions:
                        self.bot.unload_extension(f'cogs.{i[:-3]}')
                    else:
                        print(f'{Fore.YELLOW}[{datetime.now()}] [W] [RELOAD] - Is cogs.{i[:-3]} loaded?!{Style.RESET_ALL}')
                    self.bot.load_extension(f'cogs.{i[:-3]}')
            await ctx.reply('Bot Reloaded!')
            print(f'{Fore.LIGHTYELLOW_EX}[{datetime.now()}] [I] [RELOAD] - Reloaded by {ctx.author}.{Style.RESET_ALL}')
    
    @ping.error
    async def on_ping_error(self, ctx, error):
        if isinstance(error, commands.CommandOnCooldown):
            lang = self.check_lang(ctx.guild)
            if lang == "RU":
                emb = disnake.Embed(
                    title='Вы заморожены!',
                    description=f'Попробуйте выполнить команду примерно через {int(error.retry_after)} секунд(у)!',
                    color=0xff0000
                )
            elif lang == "EN":
                emb = disnake.Embed(
                    title='You are on Cooldown!',
                    description=f'Try again in about {int(error.retry_after)} second(s)!',
                    color=0xff0000
                )
            emb.set_footer(text=ctx.author, icon_url=ctx.author.avatar.url)
            await ctx.reply(embed=emb)

def setup(bot):
    bot.add_cog(Utils(bot))

cogs/guildlog.py
import disnake
from disnake.ext import commands

class guildlog(commands.Cog):
    def __init__(self,bot):
        self.bot = bot
        
    @commands.Cog.listener()
    async def on_guild_remove(self,guild):
        logchannel = await self.bot.fetch_channel(886449232486227998)
        emd = disnake.Embed(
            title=f"Вышел с сервера {guild}",
            description=f"Кол-во участников: **{guild.member_count}**",
            colour=0xff0000
        )
        await logchannel.send(embed=emd)

    @commands.Cog.listener()
    async def on_guild_join(self,guild):
        logchannel = await self.bot.fetch_channel(886449232486227998)
        emd = disnake.Embed(
            title=f"Зашел на сервер {guild}",
            description=f"Кол-во участников: **{guild.member_count}**",
            colour=0x00ff00
        )
        await logchannel.send(embed=emd)
    
def setup(bot):
    bot.add_cog(guildlog(bot))

bot.py
import disnake, json, os, random,shutil, sys
from disnake import colour
from disnake.ext import commands
from datetime import datetime
from colorama import init, Style, Back, Fore
from disnake.ext.commands.core import check
from functions import debug_info, redefine_std, backup, version_check, release_info
from copy import copy
with open("env.json", encoding="utf-8") as file: conf = json.load(file)
init()
version_check(conf["ro.python.minimal"][0], conf["ro.python.minimal"][1])
with open("token.json", encoding="utf-8") as file: token = json.load(file); TOKEN = token["vendor.token"]; del token

if conf["ro.bot.logd"]:
    redefine_std()
    print(f'{Fore.LIGHTGREEN_EX}{Style.BRIGHT}[{datetime.now()}] [D] [LOGDMN] - LogD started.{Style.RESET_ALL}')

if conf["ro.build.type"][0] == "debug": debug_info(conf)
elif conf["ro.build.type"][0] == "release": release_info(conf)

if not os.path.isdir("backup"): os.mkdir("backup")
if not os.path.isfile("credit.json"): 
    with open("credit.json", "w") as file: file.write("{}") 

backup()

# async with ctx.channel.typing():

intents = disnake.Intents.default()
intents.members = True
intents.message_content = True
bot = commands.Bot(command_prefix='sc.',owner_ids=[528606316432719908,453167201780760577],test_guilds=[883778577609412660],intents=intents)
bot.conf = copy(conf)

# ======
# ЗАПУСК
# ======

bot.remove_command("help")

with open("credit.json", "r", encoding="utf-8") as file:
    db = json.load(file)
    bot.db = db

def save():
    with open("credit.json", "w", encoding="utf-8") as file:
        json.dump(bot.db, file, indent=4)

def load():
    with open("credit.json", "r", encoding="utf-8") as file:
        bot.db = json.load(file)

def check_lang(guild):
    load()
    gid = str(guild.id)
    return bot.db[gid]["lang"]

@bot.event
async def on_command_error(ctx, error):
    blacklist = ["MissingPermissions", "MemberNotFound", "CommandNotFound", "CommandOnCooldown"] # Расизм, расия)
    if isinstance(error, disnake.ext.commands.CommandNotFound):
        err = str(error).replace("Command \"", "").replace("\" is not found", "")
        lang = check_lang(ctx.guild)
        if lang == "RU":
            emb = disnake.Embed(
                title='Команда не найдена!',
                description=f'Ну и ну! Бот не смочь найти команда "{err}"!',
                color=0xff0000
            )
        elif lang == "EN":
            emb = disnake.Embed(
                title='Command not found!',
                description=f'Well, the Bot couldn\'t find the "{err}" command!',
                color=0xff0000
            )
        emb.set_image(url='https://media.discordapp.net/attachments/883779765415337995/896458794886918225/unknown.png')
        emb.set_footer(text=ctx.author, icon_url=ctx.author.avatar.url)

        await ctx.reply(embed=emb)
    if type(error).__name__ in blacklist: return
    print(f'{Fore.RED}[{datetime.now()}] [C] [ERRMSG] - Error Raised! More info below:{Style.RESET_ALL}')
    print(f"{Fore.LIGHTRED_EX}-> {error}{Style.RESET_ALL}")

@bot.event
async def on_ready():
    stream = disnake.Game('patching for / cmds')
    await bot.change_presence(status=disnake.Status.idle, activity=stream)
    print(f'{Fore.GREEN}[{datetime.now()}] [I] [CLIENT] - Launched.{Style.RESET_ALL}')
    for i in os.listdir("cogs"):
        if os.path.isfile(os.path.join("cogs", i)):
            if not f'cogs.{i[:-3]}' in bot.extensions:
                bot.load_extension(f'cogs.{i[:-3]}')

@bot.command()
async def guildcount(ctx):
    await ctx.reply(len(bot.guilds))

# =======
# КОМАНДЫ
# =======

def ownercheck(id):
    if id in [528606316432719908,453167201780760577]:
        return True
    else:
        return False

bot.ownercheck = ownercheck

# ТОКЕН
bot.run(TOKEN)